File: modules/portfolio-tracker/src/extractors/ibkr_parser.py
from lxml import etree


IBKR_NS = {"ns": "http://www.interactivebrokers.com/flex/statement"}

TRADE_TYPE_MAP = {
    "BUY": "Buy",
    "SELL": "Sell",
    "BUY (CA)": "Buy",
    "SELL (CA)": "Sell",
}

CORP_ACTION_MAP = {
    "DIVIDEND": "Dividend",
    "DIVIDEND_REINVEST": "Dividend",
    "STOCK_SPLIT": "Buy",
}


def parse_ibkr_flex_query(xml_content: str) -> list[dict]:
    try:
        root = etree.fromstring(xml_content.encode("utf-8"))
    except etree.XMLSyntaxError as e:
        raise ValueError(f"Invalid IBKR flex query XML: {e}")

    transactions: list[dict] = []

    trades_elem = root.find(".//ns:Trades", IBKR_NS)
    if trades_elem is not None:
        for trade in trades_elem.findall("ns:Trade", IBKR_NS):
            txn = _parse_trade(trade)
            if txn:
                transactions.append(txn)

    cash_elem = root.find(".//ns:CashTransactions", IBKR_NS)
    if cash_elem is not None:
        for ct in cash_elem.findall("ns:CashTransaction", IBKR_NS):
            txn = _parse_cash_transaction(ct)
            if txn:
                transactions.append(txn)

    corp_elem = root.find(".//ns:CorporateActions", IBKR_NS)
    if corp_elem is not None:
        for ca in corp_elem.findall("ns:CorporateAction", IBKR_NS):
            txn = _parse_corporate_action(ca)
            if txn:
                transactions.append(txn)

    return transactions


def _get_text(elem, tag: str) -> str:
    val = elem.get(tag)
    if val is not None:
        return val.strip()
    child = elem.find(f"ns:{tag}", IBKR_NS)
    if child is not None:
        return (child.text or "").strip()
    return ""


def _get_float(elem, tag: str) -> float:
    val = _get_text(elem, tag)
    if not val:
        return 0.0
    try:
        return float(val.replace(",", ""))
    except (ValueError, TypeError):
        return 0.0


def _get_date(elem, tag: str) -> str:
    val = _get_text(elem, tag)
    if val and len(val) >= 10:
        return val[:10]
    return val


def _parse_trade(trade) -> dict | None:
    symbol = _get_text(trade, "symbol")
    isin = _get_text(trade, "isin")
    if not symbol or symbol == "":
        return None

    trade_type_raw = _get_text(trade, "buySell")
    trade_type = TRADE_TYPE_MAP.get(trade_type_raw, trade_type_raw)

    date = _get_date(trade, "tradeDate") or _get_date(trade, "dateTime")
    quantity = abs(_get_float(trade, "quantity"))
    price = _get_float(trade, "tradePrice") or _get_float(trade, "price")
    currency = _get_text(trade, "currency") or _get_text(trade, "currencyCode")
    fees = _get_float(trade, "ibCommission") + _get_float(trade, "taxes")
    taxes = 0.0
    description = _get_text(trade, "description") or f"{symbol} {trade_type_raw}"

    amount = _get_float(trade, "proceeds")
    if amount < 0 and trade_type == "Buy":
        amount = abs(amount)

    return {
        "type": trade_type,
        "symbol": symbol,
        "isin": isin,
        "description": description,
        "date": date,
        "quantity": quantity,
        "price": price,
        "currency": currency,
        "fees": fees,
        "taxes": taxes,
        "amount": amount,
        "source": "IBKR_Trade",
    }


def _parse_cash_transaction(ct) -> dict | None:
    ctype = _get_text(ct, "type")
    if not ctype:
        return None

    type_map = {
        "Dividends": "Dividend",
        "Dividend": "Dividend",
        "Withholding Tax": "Tax",
        "Withholding": "Tax",
        "Deposits/Withdrawals": "Deposit",
        "Deposit": "Deposit",
        "Withdrawal": "Withdrawal",
        "Interest": "Interest",
        "Broker Interest": "Interest",
        "Other Fees": "Fee",
        "Fees": "Fee",
    }

    txn_type = type_map.get(ctype, ctype)
    date = _get_date(ct, "dateTime") or _get_date(ct, "reportDate")
    currency = _get_text(ct, "currency")
    amount = _get_float(ct, "amount")
    description = _get_text(ct, "description") or ctype
    symbol = _get_text(ct, "symbol")

    if amount < 0 and txn_type in ("Deposit", "Dividend", "Interest"):
        amount = abs(amount)
    elif amount > 0 and txn_type in ("Withdrawal", "Fee", "Tax"):
        amount = -abs(amount)

    return {
        "type": txn_type,
        "symbol": symbol,
        "isin": "",
        "description": description,
        "date": date,
        "quantity": 0,
        "price": abs(amount) if amount else 0,
        "currency": currency,
        "fees": 0.0,
        "taxes": 0.0,
        "amount": amount,
        "source": "IBKR_Cash",
    }


def _parse_corporate_action(ca) -> dict | None:
    ca_type = _get_text(ca, "type")
    if not ca_type:
        return None

    mapped_type = CORP_ACTION_MAP.get(ca_type, "Dividend")
    symbol = _get_text(ca, "symbol")
    date = _get_date(ca, "dateTime") or _get_date(ca, "reportDate")
    currency = _get_text(ca, "currency")
    description = _get_text(ca, "description") or f"{ca_type}: {symbol}"
    quantity = _get_float(ca, "quantity")

    return {
        "type": mapped_type,
        "symbol": symbol,
        "isin": "",
        "description": description,
        "date": date,
        "quantity": quantity,
        "price": 0.0,
        "currency": currency,
        "fees": 0.0,
        "taxes": 0.0,
        "amount": 0.0,
        "source": "IBKR_CA",
    }

File: modules/portfolio-tracker/src/ibkr_parser.js
/**
 * IBKR Flex Query XML parser.
 * Ported 1:1 from src/extractors/ibkr_parser.py
 *
 * Parses IBKR flex query XML into structured transaction objects.
 * Handles namespace-aware and namespace-less XML.
 * Parses: Trades, CashTransactions, CorporateActions.
 */

const IBKR_NS = "http://www.interactivebrokers.com/flex/statement";

const TRADE_TYPE_MAP = {
    BUY: "Buy",
    SELL: "Sell",
    "BUY (CA)": "Buy",
    "SELL (CA)": "Sell",
};

const CORP_ACTION_MAP = {
    DIVIDEND: "Dividend",
    DIVIDEND_REINVEST: "Dividend",
    STOCK_SPLIT: "Buy",
};

const CASH_TYPE_MAP = {
    Dividends: "Dividend",
    Dividend: "Dividend",
    "Withholding Tax": "Tax",
    Withholding: "Tax",
    "Deposits/Withdrawals": "Deposit",
    Deposit: "Deposit",
    Withdrawal: "Withdrawal",
    Interest: "Interest",
    "Broker Interest": "Interest",
    "Other Fees": "Fee",
    Fees: "Fee",
};

/**
 * Get text content of a child element or attribute.
 * Tries namespace-qualified first, then unqualified.
 */
function getTextNS($, el, tag) {
    // Try attribute
    const attr = $(el).attr(tag);
    if (attr !== undefined && attr !== null) return attr.trim();

    // Try ns:Tag
    let child = $(el).find(`ns\\:${tag}`).first();
    if (child.length > 0) return child.text().trim();

    // Try Tag (no namespace)
    child = $(el).find(tag).first();
    if (child.length > 0) return child.text().trim();

    return "";
}

function getFloatNS($, el, tag) {
    const val = getTextNS($, el, tag);
    if (!val) return 0;
    const cleaned = val.replace(/,/g, "");
    const parsed = parseFloat(cleaned);
    return isNaN(parsed) ? 0 : parsed;
}

function getDateNS($, el, tag) {
    const val = getTextNS($, el, tag);
    if (val && val.length >= 10) return val.slice(0, 10);
    return val;
}

function parseTrade($, trade) {
    const symbol = getTextNS($, trade, "symbol");
    const isin = getTextNS($, trade, "isin");
    if (!symbol) return null;

    const tradeTypeRaw = getTextNS($, trade, "buySell");
    const tradeType = TRADE_TYPE_MAP[tradeTypeRaw] || tradeTypeRaw;

    const date =
        getDateNS($, trade, "tradeDate") || getDateNS($, trade, "dateTime");
    const quantity = Math.abs(getFloatNS($, trade, "quantity"));
    const price =
        getFloatNS($, trade, "tradePrice") || getFloatNS($, trade, "price");
    const currency =
        getTextNS($, trade, "currency") || getTextNS($, trade, "currencyCode");
    const ibCommission = getFloatNS($, trade, "ibCommission");
    const xmlTaxes = getFloatNS($, trade, "taxes");
    const fees = ibCommission + xmlTaxes;
    const taxes = 0; // commission already included in fees
    const description =
        getTextNS($, trade, "description") || `${symbol} ${tradeTypeRaw}`;

    let amount = getFloatNS($, trade, "proceeds");
    if (amount < 0 && tradeType === "Buy") {
        amount = Math.abs(amount);
    }

    return {
        type: tradeType,
        symbol,
        isin,
        description,
        date,
        quantity,
        price,
        currency,
        fees,
        taxes,
        amount,
        source: "IBKR_Trade",
    };
}

function parseCashTransaction($, ct) {
    const ctype = getTextNS($, ct, "type");
    if (!ctype) return null;

    const txnType = CASH_TYPE_MAP[ctype] || ctype;
    const date = getDateNS($, ct, "dateTime") || getDateNS($, ct, "reportDate");
    const currency = getTextNS($, ct, "currency");
    let amount = getFloatNS($, ct, "amount");
    const description = getTextNS($, ct, "description") || ctype;
    const symbol = getTextNS($, ct, "symbol");

    // Fix sign conventions
    if (
        amount < 0 &&
        (txnType === "Deposit" ||
            txnType === "Dividend" ||
            txnType === "Interest")
    ) {
        amount = Math.abs(amount);
    } else if (
        amount > 0 &&
        (txnType === "Withdrawal" || txnType === "Fee" || txnType === "Tax")
    ) {
        amount = -Math.abs(amount);
    }

    return {
        type: txnType,
        symbol,
        isin: "",
        description,
        date,
        quantity: 0,
        price: amount ? Math.abs(amount) : 0,
        currency,
        fees: 0,
        taxes: 0,
        amount,
        source: "IBKR_Cash",
    };
}

function parseCorporateAction($, ca) {
    const caType = getTextNS($, ca, "type");
    if (!caType) return null;

    const mappedType = CORP_ACTION_MAP[caType] || "Dividend";
    const symbol = getTextNS($, ca, "symbol");
    const date = getDateNS($, ca, "dateTime") || getDateNS($, ca, "reportDate");
    const currency = getTextNS($, ca, "currency");
    const description =
        getTextNS($, ca, "description") || `${caType}: ${symbol}`;
    const quantity = getFloatNS($, ca, "quantity");

    return {
        type: mappedType,
        symbol,
        isin: "",
        description,
        date,
        quantity,
        price: 0,
        currency,
        fees: 0,
        taxes: 0,
        amount: 0,
        source: "IBKR_CA",
    };
}

/**
 * Parse an IBKR Flex Query XML response and extract all transactions.
 * @param {string} xmlContent - Raw XML string from IBKR flex query
 * @returns {object[]} Array of transaction objects
 */
export async function parseIBKRFlexQuery(xmlContent) {
    // Use cheerio with xmlMode
    const cheerioMod = await import("cheerio");
    const cheerio = cheerioMod.default || cheerioMod;
    const $ = cheerio.load(xmlContent, { xmlMode: true });

    const transactions = [];

    // --- Parse Trades ---
    // Try ns:Trades first
    let tradesElem = $("Trades").first();
    if (tradesElem.length > 0) {
        tradesElem.find("Trade").each((_, trade) => {
            const txn = parseTrade($, trade);
            if (txn) transactions.push(txn);
        });
    }

    // --- Parse CashTransactions ---
    let cashElem = $("CashTransactions").first();
    if (cashElem.length > 0) {
        cashElem.find("CashTransaction").each((_, ct) => {
            const txn = parseCashTransaction($, ct);
            if (txn) transactions.push(txn);
        });
    }

    // --- Parse CorporateActions ---
    let corpElem = $("CorporateActions").first();
    if (corpElem.length > 0) {
        corpElem.find("CorporateAction").each((_, ca) => {
            const txn = parseCorporateAction($, ca);
            if (txn) transactions.push(txn);
        });
    }

    return transactions;
}
